js/impress.js

```diff
@@ -668,6 +670,334 @@
+ * can call `impress().lib.gc.teardown()` to cause all of them to be removed from DOM, so that
+ * the document is in the state it was before calling `impress().init()`.
+ *
+ * In addition to just adding elements and event listeners to the garbage collector, plugins
+ * can also register callback functions to do arbitrary cleanup upon teardown.
+ *
+ * Henrik Ingo (c) 2016
+ * MIT License
+ */
+
+( function( document, window ) {
+    "use strict";
+    var roots = [];
+    var rootsCount = 0;
+    var startingState = { roots: [] };
+
+    var libraryFactory = function( rootId ) {
+        if ( roots[ rootId ] ) {
+            return roots[ rootId ];
+        }
+
+        // Per root global variables (instance variables?)
+        var elementList = [];
+        var eventListenerList = [];
+        var callbackList = [];
+
+        recordStartingState( rootId );
+
+        // LIBRARY FUNCTIONS
+        // Below are definitions of the library functions we return at the end
+        var pushElement = function( element ) {
+            elementList.push( element );
+        };
+
+        // Convenience wrapper that combines DOM appendChild with gc.pushElement
+        var appendChild = function( parent, element ) {
+            parent.appendChild( element );
+            pushElement( element );
+        };
+
+        var pushEventListener = function( target, type, listenerFunction ) {
+            eventListenerList.push( { target:target, type:type, listener:listenerFunction } );
+        };
+
+        // Convenience wrapper that combines DOM addEventListener with gc.pushEventListener
+        var addEventListener = function( target, type, listenerFunction ) {
+            target.addEventListener( type, listenerFunction );
+            pushEventListener( target, type, listenerFunction );
+        };
+
+        // If the above utilities are not enough, plugins can add their own callback function
+        // to do arbitrary things.
+        var addCallback = function( callback ) {
+            callbackList.push( callback );
+        };
+        addCallback( function( rootId ) { resetStartingState( rootId ); } );
+
+        var teardown = function() {
+
+            // Execute the callbacks in LIFO order
+            var i; // Needed by jshint
+            for ( i = callbackList.length - 1; i >= 0; i-- ) {
+                callbackList[ i ]( rootId );
+            }
+            callbackList = [];
+            for ( i = 0; i < elementList.length; i++ ) {
+                elementList[ i ].parentElement.removeChild( elementList[ i ] );
+            }
+            elementList = [];
+            for ( i = 0; i < eventListenerList.length; i++ ) {
+                var target   = eventListenerList[ i ].target;
+                var type     = eventListenerList[ i ].type;
+                var listener = eventListenerList[ i ].listener;
+                target.removeEventListener( type, listener );
+            }
```

Comment:
Can we use `.forEach` instead of imperative loop?

Reply:
'js/impress.js' is a generated file (built from src/impress.js and other source files), so please comment on source files rather then generated file.

Reply:
And I don't mind it being a regular for loop instead of `.forEach`.
Especially that in cases like calling callbacks it's used to call them in specific order.

Reply:
> 'js/impress.js' is a generated file (built from src/impress.js and other source files), so please comment on source files rather then generated file.

Oh right, didn't realize that lol.

> Especially that in cases like calling callbacks it's used to call them in specific order.

If you are referring to the callback being executed in the inverse order, it's just a matter of calling:

```javascript
callbackList.reverse().forEach;
```

That can simplify the code and reduce the vectors of failure (less moving parts and variables like the index assignment and manual increment)

Reply:
`reverse` is reversing array in place which may not always expected result as it changes the array for later uses, which by itself is a 'moving part' to remember about.

I agree that .forEach makes code more concise, but it also introduces additional function, which adds a bit of complexity by itself:

```
elemenList.forEach(function(element) {
   element.parentElement.removeChild(element);
});
```

So it really has pros and cons and I'm fine with leaving it to @henrikingo preference and would certainly not block merging this PR on it.

Reply:
That's cool then.

Reply:
Code base has avoided forEach due to the version of buildify that was used wouldn't support it. Recently I actually disabled the minify part of buildify, so this is no longer an issue, but it is the reason existing code doesn't use forEach.

(And as explained elsewhere, these kind of code style comments are out of scope as a general category. The goal of these PRs is to merge code that already exists in my repo.)